hyperspy/drawing/markers.py

```diff
@@ -0,0 +1,676 @@
+        self.is_iterating = np.any(
+            [is_iterating(value) for key, value in self.kwargs.items()]
+        )
+        self._plot_on_signal = True
+        self.name = "Markers"
+        self.plot_marker = True
+        self.offsets_structure = [["o", "get_xlim[0]"], ["o", "get_xlim[1]"]]
+
+        # Events
+        self.events = Events()
+        self.events.closed = Event(
+            """
+            Event triggered when a marker is closed.
+
+            Arguments
+            ---------
+            marker : Marker
+                The marker that was closed.
+            """,
+            arguments=["obj"],
+        )
+        self._closing = False
+
+    def _check_iterating_kwargs(self):
+        if self.collection_class is None:
+            cc = matplotlib.collections.PathCollection
+        else:
+            cc = self.collection_class
+        # for key, value in self.kwargs.items():
+        #     if is_iterating(value) and key not in str(inspect.signature(cc.set)):
+        #         raise ValueError(f"The keyword {key} is an dynamic argument and cannot be set using the"
+        #                          f" {self.collection_class.__name__}.set function.  Another option is to"
+        #                          f" use the IterPatchCollection which provides additional control at the"
+        #                          f" cost of speed.")
+
+    def delete_index(self, keys, index):
+        """
+        Delete the index from the kwargs.
+
+        Parameters
+        ----------
+        keys: list
+            List of keys to delete from.
+        index: slice, int or array of ints
+            Indicate indices of sub-arrays to remove along the specified axis.
+        """
+        if isinstance(keys, str):
+            keys = [
+                keys,
+            ]
+        for key in keys:
+            if self.kwargs[key].dtype == object:
+                for i in np.ndindex(self.kwargs[key].shape):
+                    self.kwargs[key][i] = np.delete(self.kwargs[key][i], index, axis=0)
+            else:
+                self.kwargs[key] = np.delete(self.kwargs[key], index, axis=0)
+
+    def append_kwarg(self, keys, value):
+        """
+        Add the index from the kwargs.
+
+        Parameters
+        ----------
+        keys: list
+            List of keys to append.
+        value:
+            The value to append to the kwarg.
+        """
+        if isinstance(keys, str):
+            keys = [
+                keys,
+            ]
+        # skip if value is empty
+        if len(value) == 0:
+            return
+        for key in keys:
+            if self.kwargs[key].dtype == object:
+                for i in np.ndindex(self.kwargs[key].shape):
+                    self.kwargs[key][i] = np.append(self.kwargs[key][i], value, axis=0)
+            else:
+                self.kwargs[key] = np.append(self.kwargs[key], value, axis=0)
+
+    def _get_cache_dask_kwargs_chunk(self, indices):
```

Comment:
I am wondering if it would be worth using `functools.lru_cache` here since the output shouldn't used that much memory?

Reply:
That is definitely the right thing to do. We should create a second function like `get_current_chunk` which is cached for each chunk/key rather than for each index.  The problem is that if you  do that currently you might end up moving from chunk 1 to chunk2 an then back to a new index in chunk 1 and it will recompute the entire chunk.

The same option should be added for plotting as well but maybe just limiting the number of chunks saved to like 10 or so.

Reply:
Ideally we should have some`LazySignal` have an attribute ._to_compute and da.compute should only be called once.

For example currently if you want to load some data, find some peaks and then plot that data what will happen is.


1. Chunk is loaded
2. Chunk is plotted
3. Chunk is loaded second time
4. Peaks are found
5. Peaks are plotted

So ideally we need to build a function

`hs.compute()` which can take signals or markers and then computes/ updates the relevant information so that 

1.  Chunk is loaded
2.  Peaks are found
3. Chunk is plotted
4. Peaks are plotted.


Or more ideally:

0. Cache is checked to see if chunk/peaks are already computed
2.  Chunk is loaded
3.  Peaks are found
4. Chunk is plotted
5. Peaks are plotted

Reply:
Can this be a later PR though... For now lets just make things work...